test/extended/ha/ha.go

```diff
@@ -0,0 +1,258 @@
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("ha: Election of OpenShift controllers", func() {
+	defer g.GinkgoRecover()
+	var (
+		podConfig  = exutil.FixturePath("fixtures", "hello-world-pod.json")
+		reExitCode = regexp.MustCompile(`(?i)exitcode:(\d+)`)
+	)
+
+	// Run a controller manager with one intitial controllers instance.
+	// Running controllers are needed for NewCLI to run.
+	// FIXME: seems like OutputDir is unset at this moment (execution of the first test)
+	m, _err := exutil.NewControllersManager(exutil.MasterConfigPath(), 0, 8444, exutil.TestContext.OutputDir)
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	// FIXME: Controllers instance immediatelly terminates after acquiring its lease without following sleep
+	time.Sleep(1 * time.Second)
+
+	// One controllers instance needs to be running before CLI can be initialized
+	_, _err = m.StartNewInstance()
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	_, _, _err = m.WaitForActive(time.Second * 2)
+	if _err != nil {
+		g.Fail(_err.Error())
+	}
+
+	oc := exutil.NewCLI("controller-leases", exutil.KubeConfigPath())
+
+	// WaitForPod blocks until given pod terminates.
+	// FIXME: this is probably too complicated
+	// FIXME: consider using oc logs directly
+	WaitForPod := func(podName string, timeout time.Duration) {
+		start := time.Now()
+		cmd := exec.Command("oc", "get", "-n", oc.Namespace(), "--watch", "--no-headers", "pod", podName)
+		stdout, err := cmd.StdoutPipe()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		stderr, err := cmd.StderrPipe()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		ec := make(chan error)
+		err = cmd.Start()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// log stderr
+		go func() {
+			io.Copy(g.GinkgoWriter, stderr)
+		}()
+
+		// log stdout and look for exited state
+		go func() {
+			bufReader := bufio.NewReader(stdout)
+			for {
+				line, err := bufReader.ReadString('\n')
+				if err != nil {
+					ec <- err
+					return
+				}
+				g.GinkgoWriter.Write([]byte(fmt.Sprintf("`oc get --watch` output: %s", line)))
+				submatch := reExitCode.FindStringSubmatch(line)
+				if len(submatch) > 1 {
+					if submatch[1] != "0" {
+						ec <- fmt.Errorf("Pod terminated with unexpected exit code (%s)", submatch[1])
+					} else {
+						ec <- nil
+					}
+					return
+				}
+			}
+		}()
+
+		select {
+		case err := <-ec:
+			if err != nil && err != io.EOF {
+				o.Expect(err).NotTo(o.HaveOccurred())
+			}
+			if err == nil {
+				cmd.Process.Kill()
+				cmd.Wait()
+			} else {
+				o.Expect(cmd.Wait()).NotTo(o.HaveOccurred())
+			}
+		case <-time.After(start.Add(timeout).Sub(time.Now())):
+			g.Fail(fmt.Sprintf("Timeout (%s) occurred while waiting for pod's exit", timeout.String()))
+		}
+	}
+
+	StartPod := func() {
```

Comment:
Explain what this is for.  Is this just a function to ensure that the server still works?  If so, I'd like to see a deploymentconfig.